src/pci/smccc/smccc.c

```diff
@@ -193,6 +225,335 @@ static int smccc_pci_version_test(fwts_framework *fw)
 	return FWTS_OK;
 }
 
+
+/*
+ *  smccc_version_test()
+ *	test SMCCC function VERSION
+ */
+static int smccc_version_test(fwts_framework *fw)
+{
+	int ret;
+	struct smccc_test_arg arg = { };
+
+	arg.size = sizeof(arg);
+	arg.w[0] = ARM_SMCCC_VERSION;
+
+	ret = ioctl(smccc_fd, SMCCC_TEST_VERSION_FUNCTION, &arg);
+	if (ret < 0) {
+		fwts_log_error(fw, "SMCCC test driver ioctl SMCCC_TEST_VERSION "
+			"failed, errno=%d (%s)\n", errno, strerror(errno));
+		return FWTS_ERROR;
+	}
+	if (smccc_pci_conduit_check(fw, &arg) != FWTS_OK)
+		return FWTS_ERROR;
+
+	fwts_log_info_verbatim(fw, "  SMCCC conduit type: 0x%x ('%s')",
+		arg.conduit, smccc_pci_conduit_name(&arg));
+
+	int32_t return_value = arg.w[0];
+	
+	//NOT_SUPPORTED is treated as v1.0
+	if(return_value == ARM_SMCCC_RET_NOT_SUPPORTED)
+	{
+		smccc_major_version = 1;
+		smccc_minor_version = 0;
+		fwts_log_info_verbatim(fw, "Arm SMCCC: Major version: %d, Minor version: %d",
+			smccc_major_version, smccc_minor_version);
+	}
+	else
+	{
+		//Bit 31 should not be 1. Bits 30...0 should be non 0 
+		if((arg.w[0] >> 31) & 0x1 || ((arg.w[0] & 0x7FFFFFFF) == 0))
+		{
+			fwts_log_error(fw, "SMCCC_TEST_VERSION return value is invalid." 
+				"arg.w[0] = 0x%8.8" PRIx32,
+				arg.w[0]);
+			//fwts_failed(fw, "Arm SMCCC_VERSION failed");
+			return FWTS_ERROR;
+		}
+
+		smccc_major_version = arg.w[0] >> 16;
+		smccc_minor_version = arg.w[0] & 0x0000ffff;
+		fwts_log_info_verbatim(fw, "Arm SMCCC: Major version: %d, Minor version: %d",
+			smccc_major_version, smccc_minor_version);
+
+	}
+
+	fwts_passed(fw, "Arm SMCCC_VERSION test passed");
+
+	return FWTS_OK;
+}
+
+/*
+ * smccc_arch_features_bbr_check()
+ * Asserts based on the Arm BBR specification rules in
+ * Section 4.1. SMCCC architecture call requirements 
+ * This function can be run only if SBBR or EBBR flag is set
+ * while running FWTS
+ */
+
+ static int smccc_arch_features_bbr_check(fwts_framework *fw)
+ {
+	uint32_t version = (smccc_major_version << 16) + smccc_minor_version;
+	bool bbr_check_failed = false; 
+	uint16_t assert_failed_count = 0;
+	int i=0;
+
+	fwts_log_info_verbatim(fw, "SMCCC Arm BBR check - the detected SMCCC version is %d.%d",
+		smccc_major_version, smccc_minor_version);
+
+	for (i=0; i<FWTS_ARRAY_SIZE(SMCCC_func_id_list); i++)
+	{
+		if(SMCCC_func_id_list[i].implemented == false)
+		{
+			switch (SMCCC_func_id_list[i].SMCCC_func_id)
+			{
+				case ARM_SMCCC_VERSION:
+				case ARM_SMCCC_ARCH_FEATURES:
+				{
+					if(version >= SMCCC_VERSION_1_1)
+					{
+						bbr_check_failed = true;
+					}
+				}
+				break; 
+				
+				case ARM_SMCCC_ARCH_SOC_ID:
+				{
+					if(version >= SMCCC_VERSION_1_2)
+					{
+						bbr_check_failed = true;
+					}
+				}
+				break;
+
+				default:
+				{
+					fwts_log_info_verbatim(fw, "Info: Function_id: 0x%8.8" PRIx32 " (%s) is not Implemented."
+						"\nHowever this function is not \"Required\" for this version of SMCCC as per the Arm BBR specification",
```

Comment:
"\n" is necessary?
Maybe you can use another  fwts_log_info_verbatim for another line.

Reply:
Yes I intended to print on another line. Will follow your advice and have another fwts_log_info_verbatim